api/applications.go

```diff
@@ -0,0 +1,221 @@
+	AppID               string
+	AppObjectID         string
+	Description         string
+	DisplayName         string
+	PasswordCredentials []PasswordCredential
+}
+
+type PasswordCredential struct {
+	DisplayName string
+	EndDate     time.Time
+	KeyID       string
+	SecretText  string
+}
+
+// NewMSGraphApplicationClient returns a new MSGraphClient configured to interact with
+// the Microsoft Graph API. It can be configured to target alternative national cloud
+// deployments via graphURI. For details on the client configuration see
+// https://learn.microsoft.com/en-us/graph/sdks/national-clouds
+func NewMSGraphApplicationClient(graphURI string, creds azcore.TokenCredential) (*MSGraphClient, error) {
+	scopes := []string{
+		fmt.Sprintf("%s/.default", graphURI),
+	}
+
+	authProvider, err := auth.NewAzureIdentityAuthenticationProviderWithScopes(creds, scopes)
+	if err != nil {
+		return nil, err
+	}
+
+	adapter, err := msgraphsdkgo.NewGraphRequestAdapter(authProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	adapter.SetBaseUrl(fmt.Sprintf("%s/v1.0", graphURI))
+	client := msgraphsdkgo.NewGraphServiceClient(adapter)
+
+	ac := &MSGraphClient{
+		client: client,
+	}
+	return ac, nil
+}
+
+func (c *MSGraphClient) GetApplication(ctx context.Context, clientID string) (Application, error) {
+	filter := fmt.Sprintf("appId eq '%s'", clientID)
+	req := applications.ApplicationsRequestBuilderGetRequestConfiguration{
+		QueryParameters: &applications.ApplicationsRequestBuilderGetQueryParameters{
+			Filter: &filter,
+		},
+	}
+
+	resp, err := c.client.Applications().Get(ctx, &req)
+	if err != nil {
+		return Application{}, err
+	}
+
+	apps := resp.GetValue()
+	if len(apps) == 0 {
+		return Application{}, fmt.Errorf("no application found")
+	}
+	if len(apps) > 1 {
+		return Application{}, fmt.Errorf("multiple applications found - double check your client_id")
+	}
+
+	app := apps[0]
+
+	application := Application{
+		ID:                  *app.GetId(),
+		AppID:               *app.GetAppId(),
+		AppObjectID:         *app.GetId(),
+		Description:         *app.GetDescription(),
+		DisplayName:         *app.GetDisplayName(),
+		PasswordCredentials: getPasswordCredentialsForApplication(app),
+	}
+
+	return application, nil
+}
+
+func (c *MSGraphClient) ListApplications(ctx context.Context, filter string) ([]Application, error) {
+
+	req := &applications.ApplicationsRequestBuilderGetQueryParameters{
+		Filter: &filter,
+	}
+	configuration := &applications.ApplicationsRequestBuilderGetRequestConfiguration{
+		QueryParameters: req,
+	}
+	resp, err := c.client.Applications().Get(ctx, configuration)
+	if err != nil {
+		return nil, err
+	}
+
+	var apps []Application
+	for _, app := range resp.GetValue() {
+		apps = append(apps, Application{
+			ID:                  *app.GetId(),
+			AppID:               *app.GetAppId(),
+			AppObjectID:         *app.GetId(),
+			Description:         *app.GetDescription(),
+			DisplayName:         *app.GetDisplayName(),
```

Comment:
If any of these are nil we will get a panic trying to dereference them. I got a panic running `make testacc` on `*app.GetDescription()`. Maybe we should add thin wrappers functions for these that will return zero values if the dereference is nil? Also, I don't see any reason to keep store `ID`, `Description`, and `DisplayName`. From what I can tell we only make use of `AppID` and `AppObjectID`. So I think removing the unused fields could make this work simpler.

Reply:
Ah that is a good point and wrappers are a great idea. Added thin wrapper helpers to return zero values in case the responses from MS Graph API are `nil` values. Also cut down the structs `Application` and `PasswordCredential` to only required fields. Thanks for the great points!